robot_skills/test/test_collect_arm_requirements.py

```diff
@@ -0,0 +1,118 @@
+                               "required_trajectories": ["prepare_place"], }
+
+    def __init__(self, robot):
+        smach.State.__init__(self, outcomes=['succeeded'])
+
+        assert robot.get_arm(
+            **self.REQUIRED_ARM_PROPERTIES), "None of the available arms meets all this class's" \
+                                             "requirements: {}".format(self.REQUIRED_ARM_PROPERTIES)
+
+    @staticmethod
+    def execute():
+        return 'succeeded'
+
+
+class ThirdState(smach.State):
+    REQUIRED_ARM_PROPERTIES = {"required_gripper_types": [GripperTypes.PARALLEL],
+                               "required_trajectories": ["wave"], }
+
+    def __init__(self, robot):
+        smach.State.__init__(self, outcomes=['succeeded'])
+
+        assert robot.get_arm(
+            **self.REQUIRED_ARM_PROPERTIES), "None of the available arms meets all this class's" \
+                                             "requirements: {}".format(self.REQUIRED_ARM_PROPERTIES)
+
+    @staticmethod
+    def execute():
+        return 'succeeded'
+
+
+class FirstStateMachine(smach.StateMachine):
+    def __init__(self, robot):
+
+        smach.StateMachine.__init__(self, outcomes=['succeeded'])
+
+        with self:
+            smach.StateMachine.add('FOO', FirstState(robot), transitions={'succeeded': 'BAR'})
+            smach.StateMachine.add('BAR', SecondState(robot), transitions={'succeeded': 'succeeded'})
+
+
+class SecondStateMachine(smach.StateMachine):
+    def __init__(self, robot):
+
+        smach.StateMachine.__init__(self, outcomes=['succeeded'])
+
+        with self:
+            smach.StateMachine.add('FOO', ThirdState(robot), transitions={'succeeded': 'BAR'})
+            smach.StateMachine.add('BAR', FirstStateMachine(robot), transitions={'succeeded': 'succeeded'})
+
+
+def compare_lists_in_dicts(reference, result):
+    """ Compares a reference dict with the result as produced by the system
+
+    :param reference: Ground truth dict (strings as keys and lists as values)
+    :param result: Dict according to the system
+    :return: reference and result dicts containing the non overlapping key, value pairs
+    """
+    for k in reference.keys():
+        if k in result.keys():
+            if len(reference[k]) == len(result[k]) and set(reference[k]) == set(result[k]):
+                del reference[k], result[k]
+
+
+class TestCollectArmRequirements(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.robot = get_robot('mockbot')
+
+    def test_collect_arm_requirements(self):
+        requirements = collect_arm_requirements(FirstStateMachine(self.robot))
+        reference = {'required_trajectories': ['prepare_grasp', 'prepare_place'],
+                     'required_gripper_types': ['pseudo-gripper-type-any-grasping-will-do',
+                                                'gripper-type-pinch']}
+        compare_lists_in_dicts(reference, requirements)
```

Comment:
If I understand correctly then this function deletes the matching pairs between reference and requirements. So if everything works properly, both should be empty at the end correct?

Why did you make this test like this, rather than assert that the two are equal?

Reply:
Yes that is completely true! I started out with an assert in which the two should be equal, but then @MatthijsBurgh pointed out that this would break in the case that (for some reason) the order of the requirements changes. So to allow for this I created this test. I think it is ugly, but I couldn't think of anything better. If you have any ideas let me know!

Reply:
One possibility is as follows: for every entry in results, check that it exists in reference. And following that for every entry in reference check that it exists in results. You may also look into comparing dicts in python. 
Make sure that in the error message you specify: complete results and reference and the difference which caused the error.

Reply:
That could indeed be an option, but why would you choose that over this? TBH I think this is clearer and faster, but maybe not cleaner..... @PetervDooren